File: python_src/cartridge_sdk/ui/list_view.py
"""Scrollable list with cursor selection and card-style items."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, TYPE_CHECKING

import pygame

from cartridge_sdk.input import Button, InputEvent
from cartridge_sdk.screen import _get_font
from cartridge_sdk.ui.widget import Widget

if TYPE_CHECKING:
    from cartridge_sdk.screen import Screen
    from cartridge_sdk.theme import Theme

PAD_X = 10
PAD_Y = 6
CARD_MARGIN = 3
CARD_RADIUS = 6


@dataclass
class ListItem:
    """An item in a ListView."""

    id: str
    primary_text: str
    secondary_text: str = ""
    right_text: str = ""
    metadata: Any = None


class ListView(Widget):
    """Scrollable list with card-style items and selection callback."""

    def __init__(
        self,
        items: list[ListItem] | None = None,
        on_select: Callable[[ListItem], None] | None = None,
        item_height: int = 48,
        render_item: Callable | None = None,
    ) -> None:
        self.items: list[ListItem] = items or []
        self.cursor: int = 0
        self.on_select = on_select
        self.item_height = item_height
        self.render_item = render_item

    def handle_input(self, event: InputEvent) -> bool:
        if not self.items:
            return False
        if event.action not in ("press", "repeat"):
            return False

        if event.button == Button.DPAD_UP:
            self.cursor = max(0, self.cursor - 1)
            return True
        if event.button == Button.DPAD_DOWN:
            self.cursor = min(len(self.items) - 1, self.cursor + 1)
            return True
        if event.button == Button.L2:
            visible = max(1, self._last_visible_count)
            self.cursor = max(0, self.cursor - visible)
            return True
        if event.button == Button.R2:
            visible = max(1, self._last_visible_count)
            self.cursor = min(len(self.items) - 1, self.cursor + visible)
            return True
        if event.button == Button.A and self.on_select:
            self.on_select(self.items[self.cursor])
            return True

        return False

    _last_visible_count: int = 10

    def draw(self, surface: pygame.Surface, rect: pygame.Rect, theme: Theme) -> None:
        n = len(self.items)
        visible_count = max(1, rect.height // self.item_height)
        self._last_visible_count = visible_count

        # Background
        pygame.draw.rect(surface, theme.bg, rect)

        if n == 0:
            font = _get_font("mono", 14)
            msg = font.render("No items", True, theme.text_dim)
            mx = rect.x + (rect.width - msg.get_width()) // 2
            my = rect.y + rect.height // 2 - msg.get_height() // 2
            surface.blit(msg, (mx, my))
            return

        # Clamp cursor
        self.cursor = max(0, min(self.cursor, n - 1))

        # Sliding window
        window_start = _window_start(self.cursor, n, visible_count)
        window_end = min(window_start + visible_count, n)

        clip = surface.get_clip()
        surface.set_clip(rect)

        y = rect.y
        for i in range(window_start, window_end):
            item = self.items[i]
            is_selected = i == self.cursor

            if self.render_item:
                self.render_item(surface, rect, item, is_selected, y, self.item_height, theme)
            else:
                self._draw_default_item(surface, rect, item, is_selected, y, theme)

            y += self.item_height

        surface.set_clip(clip)

        # Scroll indicator
        if n > visible_count:
            _draw_scroll_indicator(surface, rect, self.cursor, n, theme)

    def _draw_default_item(
        self,
        surface: pygame.Surface,
        rect: pygame.Rect,
        item: ListItem,
        is_selected: bool,
        y: int,
        theme: Theme,
    ) -> None:
        font_primary = _get_font("mono", 15)
        font_secondary = _get_font("mono", 12)
        font_right = _get_font("mono", 13)

        card_x = rect.x + CARD_MARGIN
        card_w = rect.width - CARD_MARGIN * 2 - 8
        card_h = self.item_height - CARD_MARGIN

        # Card background
        card_rect = pygame.Rect(card_x, y + 1, card_w, card_h)
        if is_selected:
            pygame.draw.rect(surface, theme.card_highlight, card_rect, border_radius=CARD_RADIUS)
            pygame.draw.rect(surface, theme.accent, card_rect, 1, border_radius=CARD_RADIUS)
        else:
            pygame.draw.rect(surface, theme.card_bg, card_rect, border_radius=CARD_RADIUS)

        max_text_w = card_w - PAD_X * 2 - 70

        # Primary text
        text_x = card_x + PAD_X + 4
        primary = item.primary_text
        pw = font_primary.size(primary)[0]
        if pw > max_text_w:
            while len(primary) > 1 and font_primary.size(primary + "..")[0] > max_text_w:
                primary = primary[:-1]
            primary += ".."

        primary_color = theme.text if is_selected else theme.text_dim
        rendered = font_primary.render(primary, True, primary_color)
        text_y = y + 5 if item.secondary_text else y + (card_h - rendered.get_height()) // 2
        surface.blit(rendered, (text_x, text_y))

        # Secondary text
        if item.secondary_text:
            sec = font_secondary.render(item.secondary_text, True, theme.text_dim)
            surface.blit(sec, (text_x, text_y + font_primary.get_linesize() + 1))

        # Right-aligned text
        if item.right_text:
            right_surf = font_right.render(item.right_text, True, theme.text_dim)
            rx = card_x + card_w - PAD_X - right_surf.get_width()
            ry = y + (card_h - right_surf.get_height()) // 2 + 1
            surface.blit(right_surf, (rx, ry))


def _window_start(cursor: int, total: int, visible: int) -> int:
    if total <= visible:
        return 0
    start = cursor - 1
    start = max(0, start)
    start = min(start, total - visible)
    return start


def _draw_scroll_indicator(
    surface: pygame.Surface,
    rect: pygame.Rect,
    cursor: int,
    total: int,
    theme: Theme,
) -> None:
    ind_x = rect.right - 5
    bar_top = rect.y + PAD_Y
    bar_h = rect.height - PAD_Y * 2

    if total <= 1 or bar_h <= 0:
        return

    # Track
    track_color = (
        min(theme.border[0] + 10, 255),
        min(theme.border[1] + 10, 255),
        min(theme.border[2] + 10, 255),
    )
    pygame.draw.line(surface, track_color, (ind_x, bar_top), (ind_x, bar_top + bar_h))

    # Thumb
    thumb_h = max(8, bar_h // total)
    progress = cursor / (total - 1)
    thumb_y = bar_top + int((bar_h - thumb_h) * progress)
    pygame.draw.rect(surface, theme.text_dim, (ind_x - 1, thumb_y, 3, thumb_h), border_radius=1)

File: crates/cartridge-core/src/ui/list_view.rs
//! Scrollable focus list with D-pad navigation, custom row heights, L1/R1 page, scroll indicator.

use sdl2::pixels::Color;
use sdl2::rect::Rect;

use crate::input::{Button, InputAction, InputEvent};
use crate::screen::Screen;

const PAD_X: i32 = 10;
const PAD_Y: i32 = 6;
const CARD_MARGIN: i32 = 3;
const CARD_RADIUS: i16 = 6;

/// An item in a ListView.
#[derive(Clone, Debug)]
pub struct ListItem {
    pub id: String,
    pub primary_text: String,
    pub secondary_text: String,
    pub right_text: String,
    /// Application-defined tag for custom rendering or data association.
    pub tag: i64,
}

impl ListItem {
    pub fn new(id: &str, primary: &str) -> Self {
        Self {
            id: id.to_string(),
            primary_text: primary.to_string(),
            secondary_text: String::new(),
            right_text: String::new(),
            tag: 0,
        }
    }

    pub fn with_secondary(mut self, text: &str) -> Self {
        self.secondary_text = text.to_string();
        self
    }

    pub fn with_right(mut self, text: &str) -> Self {
        self.right_text = text.to_string();
        self
    }

    pub fn with_tag(mut self, tag: i64) -> Self {
        self.tag = tag;
        self
    }
}

/// Scrollable list with card-style items and D-pad navigation.
///
/// Default item height is 48px; set `item_height` to 32 for compact mode.
/// D-pad Up/Down moves cursor. L1/R1 pages. A selects.
pub struct ListView {
    pub items: Vec<ListItem>,
    pub cursor: usize,
    pub item_height: i32,
    last_visible_count: usize,
}

impl ListView {
    pub fn new(items: Vec<ListItem>, item_height: i32) -> Self {
        Self {
            items,
            cursor: 0,
            item_height,
            last_visible_count: 10,
        }
    }

    /// Create with default 48px row height.
    pub fn default_height(items: Vec<ListItem>) -> Self {
        Self::new(items, 48)
    }

    /// Create with compact 32px row height.
    pub fn compact(items: Vec<ListItem>) -> Self {
        Self::new(items, 32)
    }

    /// Get the currently focused item, if any.
    pub fn focused_item(&self) -> Option<&ListItem> {
        self.items.get(self.cursor)
    }

    /// Handle D-pad and paging input. Returns `true` if consumed.
    pub fn handle_input(&mut self, event: &InputEvent) -> bool {
        if self.items.is_empty() {
            return false;
        }
        if event.action == InputAction::Release {
            return false;
        }

        match event.button {
            Button::DpadUp => {
                if self.cursor > 0 {
                    self.cursor -= 1;
                }
                true
            }
            Button::DpadDown => {
                if self.cursor + 1 < self.items.len() {
                    self.cursor += 1;
                }
                true
            }
            Button::L1 => {
                let page = self.last_visible_count.max(1);
                self.cursor = self.cursor.saturating_sub(page);
                true
            }
            Button::R1 => {
                let page = self.last_visible_count.max(1);
                self.cursor = (self.cursor + page).min(self.items.len().saturating_sub(1));
                true
            }
            _ => false,
        }
    }

    /// Returns true if the event is an A press on the current item.
    pub fn is_select_event(&self, event: &InputEvent) -> bool {
        event.button == Button::A
            && event.action == InputAction::Press
            && !self.items.is_empty()
    }

    /// Draw the list view within `rect`.
    pub fn draw(&mut self, screen: &mut Screen, rect: Rect) {
        let n = self.items.len();
        let visible_count = (rect.height() as i32 / self.item_height).max(1) as usize;
        self.last_visible_count = visible_count;

        let theme = screen.theme;

        // Background
        screen.draw_rect(rect, Some(theme.bg), true, 0, None);

        if n == 0 {
            let text = "No items";
            let tw = screen.get_text_width(text, 14, false);
            let th = screen.get_line_height(14, false);
            let mx = rect.x() + (rect.width() as i32 - tw as i32) / 2;
            let my = rect.y() + (rect.height() as i32 - th as i32) / 2;
            screen.draw_text(text, mx, my, Some(theme.text_dim), 14, false, None);
            return;
        }

        // Clamp cursor
        self.cursor = self.cursor.min(n - 1);

        // Sliding window
        let window_start = window_start(self.cursor, n, visible_count);
        let window_end = (window_start + visible_count).min(n);

        let mut y = rect.y();
        for i in window_start..window_end {
            let item = &self.items[i];
            let is_selected = i == self.cursor;
            self.draw_default_item(screen, rect, item, is_selected, y);
            y += self.item_height;
        }

        // Scroll indicator
        if n > visible_count {
            draw_scroll_indicator(screen, rect, self.cursor, n);
        }
    }

    fn draw_default_item(
        &self,
        screen: &mut Screen,
        rect: Rect,
        item: &ListItem,
        is_selected: bool,
        y: i32,
    ) {
        let theme = screen.theme;

        let card_x = rect.x() + CARD_MARGIN;
        let card_w = (rect.width() as i32 - CARD_MARGIN * 2 - 8) as u32;
        let card_h = (self.item_height - CARD_MARGIN) as u32;

        let card_rect = Rect::new(card_x, y + 1, card_w, card_h);

        if is_selected {
            screen.draw_rect(card_rect, Some(theme.card_highlight), true, CARD_RADIUS, None);
            screen.draw_rect(card_rect, Some(theme.accent), false, CARD_RADIUS, None);
        } else {
            screen.draw_rect(card_rect, Some(theme.card_bg), true, CARD_RADIUS, None);
        }

        let max_text_w = card_w - PAD_X as u32 * 2 - 70;

        // Primary text
        let text_x = card_x + PAD_X + 4;
        let primary_color = if is_selected {
            theme.text
        } else {
            theme.text_dim
        };

        let primary_h = screen.get_line_height(15, false);
        let text_y = if !item.secondary_text.is_empty() {
            y + 5
        } else {
            y + (card_h as i32 - primary_h as i32) / 2
        };

        screen.draw_text(
            &item.primary_text,
            text_x,
            text_y,
            Some(primary_color),
            15,
            false,
            Some(max_text_w),
        );

        // Secondary text
        if !item.secondary_text.is_empty() {
            let primary_lh = screen.get_line_height(15, false);
            screen.draw_text(
                &item.secondary_text,
                text_x,
                text_y + primary_lh as i32 + 1,
                Some(theme.text_dim),
                12,
                false,
                Some(max_text_w),
            );
        }

        // Right-aligned text
        if !item.right_text.is_empty() {
            let rw = screen.get_text_width(&item.right_text, 13, false);
            let rh = screen.get_line_height(13, false);
            let rx = card_x + card_w as i32 - PAD_X - rw as i32;
            let ry = y + (card_h as i32 - rh as i32) / 2 + 1;
            screen.draw_text(
                &item.right_text,
                rx,
                ry,
                Some(theme.text_dim),
                13,
                false,
                None,
            );
        }
    }
}

fn window_start(cursor: usize, total: usize, visible: usize) -> usize {
    if total <= visible {
        return 0;
    }
    let start = cursor.saturating_sub(1);
    start.min(total - visible)
}

fn draw_scroll_indicator(screen: &mut Screen, rect: Rect, cursor: usize, total: usize) {
    let theme = screen.theme;
    let ind_x = rect.x() + rect.width() as i32 - 5;
    let bar_top = rect.y() + PAD_Y;
    let bar_h = rect.height() as i32 - PAD_Y * 2;

    if total <= 1 || bar_h <= 0 {
        return;
    }

    // Track
    let track_color = Color::RGB(
        theme.border.r.saturating_add(10),
        theme.border.g.saturating_add(10),
        theme.border.b.saturating_add(10),
    );
    screen.draw_line(
        (ind_x, bar_top),
        (ind_x, bar_top + bar_h),
        Some(track_color),
        1,
    );

    // Thumb
    let thumb_h = (bar_h / total as i32).max(8);
    let progress = cursor as f32 / (total - 1) as f32;
    let thumb_y = bar_top + ((bar_h - thumb_h) as f32 * progress) as i32;
    screen.draw_rect(
        Rect::new(ind_x - 1, thumb_y, 3, thumb_h as u32),
        Some(theme.text_dim),
        true,
        1,
        None,
    );
}
